pkg/cfn/builder/iam_helper.go

```diff
@@ -18,6 +18,39 @@ type cfnTemplate interface {
 	newResource(name string, resource gfn.Resource) *gfnt.Value
 }
 
+type customPolicy struct {
+	Name          string
+	Statements    []cft.MapOfInterfaces
+	NamedPolicies []string
+}
+
+func createWellKnownPolicies(wellKnownPolicies api.WellKnownPolicies) []customPolicy {
+	var policies []customPolicy
+	if wellKnownPolicies.ImageBuilder {
+		policies = append(policies, customPolicy{NamedPolicies: []string{iamPolicyAmazonEC2ContainerRegistryPowerUser}})
+	} else if wellKnownPolicies.AutoScaler {
+		policies = append(policies, customPolicy{Name: "PolicyAutoScaling", Statements: autoScalerStatements()})
+	} else if wellKnownPolicies.AWSLoadBalancer {
+		policies = append(policies, customPolicy{Name: "PolicyAWSLoadBalancerController", Statements: loadBalancerControllerStatements()})
+	} else if wellKnownPolicies.ExternalDNS {
+		policies = append(policies,
+			customPolicy{Name: "PolicyExternalDNSChangeSet", Statements: changeSetStatements()},
+			customPolicy{Name: "PolicyExternalDNSHostedZones", Statements: externalDNSHostedZonesStatements()},
+		)
+	} else if wellKnownPolicies.CertManager {
+		policies = append(policies,
+			customPolicy{Name: "PolicyCertManagerChangeSet", Statements: changeSetStatements()},
+			customPolicy{Name: "PolicyCertManagerGetChange", Statements: certManagerGetChangeStatements()},
+		)
+		if wellKnownPolicies.ExternalDNS {
+			policies = append(policies, customPolicy{Name: "PolicyCertManagerHostedZones", Statements: certManagerHostedZonesStatements("route53:ListHostedZones", "route53:ListTagsForResource")})
```

Comment:
This can't be hit, because we're in an `else` after `if wellKnownPolicies.ExternalDNS` above.

Reply:
Yes, missed when going from handling elements of a list of policies to handling this struct.

Reply:
I see this has been updated, but I think the logic has drifted slightly.

My understanding of the similar code in `createRole` is roughly:
* If CertManager or ExternalDNS, `changeSetStatements`
* If Certmanager, `certManagerGetChangeStatements`
* If CertManager and ExternalDNS, `certManagerHostedZonesStatements` + a couple of things ExternalDNS needs that aren't in that
* If CertManager but not ExternalDNS, `certManagerHostedZonesStatements`
* If ExternalDNS but not CertManager, `externalDNSHostedZonesStatements`.

This logic is currently
* If CertManager or ExternalDNS, `changeSetStatements`
* If CertManager, `certManagerGetChangeStatements`
* If CertManager and ExternalDNS, `certManagerHostedZonesStatements` + a couple of things ExternalDNS needs that aren't in that
* If CertManager and not ExternalDNS, `certManagerHostedZonesStatements`
* If ExternalDNS, `externalDNSHostedZonesStatements` _<==_ Applied even if CertManager.

`externalDNSHostedZonesStatements` and `certManagerHostedZonesStatements` both contain `route53:ListResourceRecordSets`, and the "couple of things" above are the other two statements from `externalDNSHostedZonesStatements`. So `externalDNSHostedZonesStatements` is a strict subset of "`certManagerHostedZonesStatements` + a couple of things ExternalDNS needs that aren't in that", so the former should only if the latter has not been applied.

To be fair, I don't know why the effort is made to avoid the overlap, unless there's an actual failure on the AWS API when a policy contains multiple statements for the same Effect/Resource/Action tuple? It just makes it harder to read and reason about.

Reply:
I've simplified that code now.

Reply:
@TBBle please do take a look over the changes if you get a chance